#Lecture: CS412 Machine Learning | Homework 5
####Instructor: Alper Özpınar
####Author: Mert Atabey Dincer
####Student ID: 20637
####Email: atabeymert@sabanciuniv.edu
####Date: 19.08.2020

In [10]:
from keras.datasets import mnist
from sklearn.neural_network import MLPClassifier

In [11]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Flatten the images

image_vector_size = 28*28

x_train = x_train.reshape(x_train.shape[0], image_vector_size)

x_test = x_test.reshape(x_test.shape[0], image_vector_size)

In [12]:
activationFunctions = ["logistic", "relu"]
runs = [1,2]

for j in runs:
  if(j == 1):
    sizes = [300,100]
  else:
    sizes = [100,50]

  for i in activationFunctions:
    clf = MLPClassifier(hidden_layer_sizes=(sizes[0],sizes[1]),activation=i,solver="sgd")
    clf.fit(x_train, y_train)
    clf.predict(x_test)

    print("\nFor sizes {} and {} with activation function {}:".format(sizes[0],sizes[1],i))
    print("Training score: {}".format(clf.score(x_train, y_train)))
    print("Test score: {}".format(clf.score(x_test, y_test)))
    

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)



For sizes 300 and 100 with activation function logistic:
Training score: 0.9901
Test score: 0.9564

For sizes 300 and 100 with activation function relu:
Training score: 0.99575
Test score: 0.9625


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)



For sizes 100 and 50 with activation function logistic:
Training score: 0.9905666666666667
Test score: 0.956


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)



For sizes 100 and 50 with activation function relu:
Training score: 0.9938666666666667
Test score: 0.9508


##Neuron size: 300 and 100
##Activation Function: logistic
######Training score: 0.990
######Test score: 0.956

---

##Neuron size: 300 and 100
##Activation Function: relu
######Training score: 0.995
######Test score: 0.962

---

##Neuron size: 100 and 50
##Activation Function: logistic
######Training score: 0.990
######Test score: 0.956

---

##Neuron size: 100 and 50
##Activation Function: relu
######Training score: 0.993
######Test score: 0.950

---

#As a result, I can say that;
 

1.   For larger neuron size, for both training and test cases relu performs better than logistic.
2.   For smaller neuron size, relu activation function performs better than logistic in terms of training, however for test case it can be seen that logistic has better result.


